### PR TITLE
Update Suppression models to php7

### PR DESCRIPTION
**Pull request by aradoje**

### CODE REVIEW

File: src/Model/Suppression/BaseResponse.php [hunk 1]
```diff
@@ -20,49 +20,29 @@
...
     private $message;
 
-    /**
-     * @param string $address
-     * @param string $message
-     */
-    private function __construct($address, $message)
+    private function __construct(string $address, string $message)
     {
         $this->address = $address;
```
Comment (Nyholm) [inline src/Model/Suppression/BaseResponse.php]: We can move this to ::create

File: src/Model/Suppression/Complaint/Complaint.php [hunk 2]
```diff
@@ -16,57 +16,33 @@
...
-     * @return Complaint
-     */
-    public static function create(array $data)
+    public static function create(array $data): self
     {
         $complaint = new self($data['address']);
 
         if (isset($data['created_at'])) {
-            $complaint->setCreatedAt(new \DateTime($data['created_at']));
+            $complaint->createdAt = new \DateTime($data['created_at']);
```
Comment (Nyholm) [inline src/Model/Suppression/Complaint/Complaint.php]: Lets use `\DateTimeImmutable`

File: src/Model/Suppression/BaseResponse.php (hunk 1 above)
Comment (aradoje) [reply src/Model/Suppression/BaseResponse.php]: What do you mean? To have empty constructor and set the properties in create method?

Comment (Nyholm) [reply src/Model/Suppression/BaseResponse.php]: Yeah, correct

File: src/Model/Suppression/BaseResponse.php [hunk 3]
```diff
@@ -20,49 +20,28 @@
...
 
-    /**
-     * @return BaseResponse
-     */
-    public static function create(array $data)
+    public static function create(array $data): self
     {
-        $address = isset($data['address']) ? $data['address'] : '';
-        $message = isset($data['message']) ? $data['message'] : '';
+        $baseResponse = new static();
```
Comment (Nyholm) [inline src/Model/Suppression/BaseResponse.php]: For consistency with other models, let's call this variable `$model`.

File: src/Model/Suppression/Bounce/Bounce.php [hunk 4]
```diff
@@ -16,105 +16,51 @@
...
-    /**
-     * @return Bounce
-     */
-    public static function create(array $data)
+    public static function create(array $data): self
     {
-        $bounce = new self($data['address']);
+        $bounce = new self();
+        $bounce->address = $data['address'];
+        $bounce->createdAt = new \DateTimeImmutable();
```
Comment (Nyholm) [inline src/Model/Suppression/Bounce/Bounce.php]: We are okey with this being null if not set from APi

File: src/Model/Suppression/Bounce/Bounce.php [hunk 5]
```diff
@@ -16,105 +16,51 @@
...
 
         if (isset($data['code'])) {
-            $bounce->setCode($data['code']);
+            $bounce->code = $data['code'];
```
Comment (Nyholm) [inline src/Model/Suppression/Bounce/Bounce.php]: Do `$model->code = $data['code'] ?? null;` and remove the if-statement

File: src/Model/Suppression/Complaint/Complaint.php [hunk 6]
```diff
@@ -16,57 +16,31 @@
...
-    /**
-     * @return Complaint
-     */
-    public static function create(array $data)
+    public static function create(array $data): self
     {
-        $complaint = new self($data['address']);
+        $complaint = new self();
+        $complaint->address = $data['address'];
+        $complaint->createdAt = new \DateTimeImmutable();
```
Comment (Nyholm) [inline src/Model/Suppression/Complaint/Complaint.php]: This line could be removed

File: src/Model/Suppression/Complaint/Complaint.php [hunk 7]
```diff
@@ -16,57 +16,31 @@
...
 
-        if (isset($data['created_at'])) {
-            $complaint->setCreatedAt(new \DateTime($data['created_at']));
-        }
+        $complaint->createdAt = isset($data['created_at']) ? new \DateTimeImmutable($data['created_at']) : new \DateTimeImmutable();
```
Comment (Nyholm) [inline src/Model/Suppression/Complaint/Complaint.php]: `$model->createdAt = isset($data['created_at']) ? new \DateTimeImmutable($data['created_at']) : null`

File: src/Model/Suppression/Bounce/Bounce.php [hunk 8]
```diff
@@ -16,105 +16,44 @@
...
-        if (isset($data['code'])) {
-            $bounce->setCode($data['code']);
-        }
-        if (isset($data['error'])) {
-            $bounce->setError($data['error']);
-        }
-        if (isset($data['created_at'])) {
-            $bounce->setCreatedAt(new \DateTime($data['created_at']));
-        }
+        $model->address = $data['address'];
```
Comment (Nyholm) [inline src/Model/Suppression/Bounce/Bounce.php]: What if `address` is not set?

File: src/Model/Suppression/Bounce/Bounce.php [hunk 9]
```diff
@@ -16,105 +16,44 @@
...
+    public function getAddress(): string
     {
         return $this->address;
     }
 
-    /**
-     * @return string
-     */
-    public function getCode()
+    public function getCode(): string
```
Comment (Nyholm) [inline src/Model/Suppression/Bounce/Bounce.php]: Must be `: ?string`

Make sure to check all of these

File: src/Model/Suppression/Bounce/Bounce.php [hunk 10]
```diff
@@ -16,105 +16,44 @@
...
-    private function setError($error)
-    {
-        $this->error = $error;
-    }
-
-    /**
-     * @return \DateTime
-     */
-    public function getCreatedAt()
+    public function getCreatedAt(): \DateTimeImmutable
```
Comment (Nyholm) [inline src/Model/Suppression/Bounce/Bounce.php]: Same here: `: ?\DateTimeImmutable`

File: src/Model/Suppression/Complaint/Complaint.php [hunk 11]
```diff
@@ -16,57 +16,30 @@
...
-    public static function create(array $data)
+    public static function create(array $data): self
     {
-        $complaint = new self($data['address']);
+        $model = new self();
 
-        if (isset($data['created_at'])) {
-            $complaint->setCreatedAt(new \DateTime($data['created_at']));
-        }
+        $model->address = $data['address'];
```
Comment (Nyholm) [inline src/Model/Suppression/Complaint/Complaint.php]: What if `address` is not set?

File: src/Model/Suppression/Complaint/Complaint.php [hunk 12]
```diff
@@ -16,57 +16,30 @@
...
+    public function getAddress(): string
     {
         return $this->address;
     }
 
-    /**
-     * @return \DateTime
-     */
-    public function getCreatedAt()
+    public function getCreatedAt(): \DateTimeImmutable
```
Comment (Nyholm) [inline src/Model/Suppression/Complaint/Complaint.php]: Nullable

File: src/Model/Suppression/Bounce/Bounce.php (hunk 8 above)
Comment (aradoje) [reply src/Model/Suppression/Bounce/Bounce.php]: thought it's ment to break, that address is mandatory. Will fix it

Comment (Nyholm) [reply src/Model/Suppression/Bounce/Bounce.php]: I dont think we should assume anything that comes from the API

**Comment by Nyholm** [comment]
This feels great:
screenshot 2019-01-22 at 16 41 54

**Pull request merged**